### PR TITLE
wasm: Add support for environment variables injection

### PR DESCRIPTION
**Pull request by mathetake**
Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

Commit Message: wasm: Add support for environment variables injection
Additional Description: wasm: Add support for environment variables injection to Wasm VMs. Envoy looks for specified env keys in the host, and if found, then inject the key-value(s) to the Wasm VMs. This is the counterpart of  https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/126, and resolves https://github.com/envoyproxy/envoy/issues/14958. 
Risk Level: low
Testing: unittest
Release Notes: 
[Optional Fixes #Issue] resolves https://github.com/envoyproxy/envoy/issues/14958

### CODE REVIEW

File: api/envoy/extensions/wasm/v3/wasm.proto [hunk 1]
```diff
@@ -92,6 +92,10 @@ message VmConfig {
   // update and do a background fetch to fill the cache, otherwise fetch the code asynchronously and enter
   // warming state.
   bool nack_on_code_cache_miss = 6;
+
+  // The keys of *Envoy's* environment variables exposed to this VM through WASI's environ_get function.
+  // Note that if a key does not exist, then it will be ignored.
+  repeated string env_keys = 7;
```
Comment (htuch) [inline api/envoy/extensions/wasm/v3/wasm.proto:98]: From an API perspective, I wonder if this belongs under some message that encapsulates the "things" we want to allow the VM access to. These might include environment variables, but maybe other proxy attributes. @PiotrSikora do you have an idea what is best here?

**Comment by htuch** [comment]
@PiotrSikora can you take a pass on this from both design and implementation angle? Thanks.

File: source/extensions/common/wasm/wasm_extension.cc [hunk 2]
```diff
@@ -58,9 +60,19 @@ WasmHandleExtensionFactory EnvoyWasm::wasmFactory() {
...
       allowed_capabilities[capability.first] = proxy_wasm::SanitizationConfig();
     }
-    auto wasm = std::make_shared<Wasm>(vm_config.runtime(), vm_config.vm_id(),
-                                       anyToBytes(vm_config.configuration()), vm_key,
-                                       allowed_capabilities, scope, cluster_manager, dispatcher);
+    // clang-format off
+    std::unordered_map<std::string, std::string> envs;
+    // clang-format on
+    for (auto key : vm_config.env_keys()) {
+      auto value = std::getenv(key.data());
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/wasm_extension.cc:67]: While we expose those values as environment variables to Proxy-Wasm plugins, the source of them doesn't necessary have to be one, and I think we should allow control plane to set both: key and value, which I don't believe is possible with this PR.

Reading values from the Envoy's environment is probably fine, but I'm not sure how useful this is in practice.

Comment (htuch) [reply source/extensions/common/wasm/wasm_extension.cc:67]: @mathetake thoughts on this?

Comment (mathetake) [reply source/extensions/common/wasm/wasm_extension.cc:67]: yeah, I now agree with @PiotrSikora here and the current api for just injecting Envoy's env to Wasm VMs is not that practical. We need to carefully design for enabling control planes to inject as well. 

That said, I would close this PR for now, and will put some design in https://github.com/envoyproxy/envoy/issues/14958

**Pull request closed without merging**